[SPARK-8975][Streaming] Adds a mechanism to send a new rate from the driver to the block generator

dragos opened the pull request:
First step for [SPARK-7398](https://issues.apache.org/jira/browse/SPARK-7398).

@tdas @huitseeker

huitseeker committed:
[SPARK-8975][Streaming] Add a mechanism to send a new rate from the driver to the block generator

huitseeker committed:
[SPARK-8975][Streaming] Adds Ratelimiter unit tests w.r.t. spark.streaming.receiver.maxRate

dragos committed:
Merge pull request #15 from huitseeker/SPARK-8975

[SPARK-8975][Streaming] Adds a mechanism to send a new rate from the driver to the block generator

dragos committed:
Add a test for the propagation of a new rate limit from driver to receivers.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

dragos commented on streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala:
```diff
@@ -537,4 +537,19 @@ trait TestSuiteBase extends SparkFunSuite with BeforeAndAfter with Logging {
+   * @param cond     A boolean that should become `true`
+   * @param timemout How many millis to wait before giving up
+   */
+  def waitUntil(cond: => Boolean, timeout: Int): Unit = {
```
This method is general enough that probably deserves a better place. I couldn't find anything similar in spark-core Utils or streaming utilities (the closest one was in BatchCounter, but won't do the same thing, and relies on wait/notify). I'm open to suggestions where to add it.

tdas replied:
We have used `eventually` for doing this. See 
https://github.com/apache/spark/blob/master/streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala#L742

No need to add another method here.

dragos replied:
cool, thanks!

SparkQA commented:
[Test build #37634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37634/consoleFull) for   PR 7471 at commit [`1e7b210`](https://github.com/apache/spark/commit/1e7b210e5f6e3600e36c9854e1887f4a03e5325a).

SparkQA commented:
[Test build #37634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37634/console) for   PR 7471 at commit [`1e7b210`](https://github.com/apache/spark/commit/1e7b210e5f6e3600e36c9854e1887f4a03e5325a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37638/consoleFull) for   PR 7471 at commit [`cd1397d`](https://github.com/apache/spark/commit/cd1397d141eda98ded62491c0f2d90a2b47e56c5).

SparkQA commented:
[Test build #37638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37638/console) for   PR 7471 at commit [`cd1397d`](https://github.com/apache/spark/commit/cd1397d141eda98ded62491c0f2d90a2b47e56c5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/Receiver.scala:
```diff
@@ -271,7 +271,7 @@ abstract class Receiver[T](val storageLevel: StorageLevel) extends Serializable
 
   /** Get the attached executor. */
-  private def executor = {
+  private[streaming] def executor = {
```
Why this change? If you want to access private members, its best to use the `PrivateTester` See
https://github.com/apache/spark/blob/master/streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala#L866

dragos replied:
I didn't know about `PrivateTester`, I'll look into it. Other Spark modules (such as core) are using these access modifiers, so I thought it's common practice.

dragos replied:
`PrivateTester` won't work here. It assumes the private method is declared in the type of the receiver. It only works if the *runtime* type of the receiver is the type that declares the private method. Not the case here, where the method is declared in `Receiver`, but the invocation is on a concrete subclass (`DummyReceiver`). Relevant code [here](https://github.com/scalatest/scalatest/blob/master/scalatest/src/main/scala/org/scalatest/PrivateMethodTester.scala#L243).

I could probably code something along those lines, but I think it's better to use qualified private. Are there any downsides to relying on language support for visibility? I doubt any user-code would declare things in `o.a.s.streaming`.

tdas replied:
Aah, I see. The reason I wanted to use PrivateTest is because marking private[streaming] make it invisible through Scala compilation, but its visibly publicly in the bytecode, and therefore through Java. And since Receivers can be written in Java, I dont want to expose that. PrivateTester wont work, but the underlying mechanism will working using reflection. 

```

scala> abstract class A { private val i = 1 }
defined class A

scala> class B extends A
defined class B

scala> B.getClass().getSuperclass()
<console>:8: error: not found: value B
              B.getClass().getSuperclass()
              ^

scala> classOf[B].getSuperclass()
res1: Class[_ >: B] = class A

scala> classOf[B].getSuperclass().getDeclaredFields()
res2: Array[java.lang.reflect.Field] = Array(private final int A.i)

scala> val f = classOf[B].getSuperclass().getDeclaredFields().head
f: java.lang.reflect.Field = private final int A.i

scala> f.setAccessible(true)

scala> f.get(new B)
res4: Object = 1


```

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -180,6 +180,12 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
     logError(s"Deregistered receiver for stream $streamId: $messageWithError")
   }
 
+  /** Update a receiver's maximum rate from an estimator's update */
```
nit: There should be no references to estimator in this scala doc. Just "Update a receiver's max ingestion rate" is good enough.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala:
```diff
@@ -0,0 +1,47 @@
+/** Testsuite for testing the network receiver behavior */
+class RateLimiterSuite extends SparkFunSuite {
+
+  test("rate limiter initializes even without a maxRate set") {
```
@dragos I had mentioned in the earlier PR that we need these test to actually send data through the rateLimiter to see whether the updated rate is taking effect. Otherwise this test to see test whether `a == 10` after setting `a = 10` is nice but not very useful.

dragos replied:
We could add an integration test. On the other hand, I personally don't see great use in testing Guava's code, I assume we take the same stance towards other libraries too. So, my assumption here is that *if* the update reaches the receiver, Guava's RateLimiter is working just fine.

tdas replied:
I was not sure whether the Guava rate limiter behaves in the desired way when the rate is updated. There are several ways of implementing a updateable rate limiter, and I wanted to understand this's one behavior, and cover it in tests. However, I just digged deeper in Guava RateLimiter code, and realized that there are fundamentally bigger problems in using the Guava rate limiter, even for stable rate. It works effectively using token bucket philosophy, which allows a large burst of data to be allowed, if there has been no data for while. We DONT want that. This means that we need to implement our own rate limiter (which was the case, until someone replaced my implementation with Guava). Anyways, thats not something outside the scope of this PR. So I am okay not testing the rate limiting. 

So please revert ReceiverSuite, and remove the block manager unit tests in this testsuite. Basically, revert it to the previous iteration (sorry :) )
-

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala:
```diff
@@ -34,12 +36,28 @@ import org.apache.spark.{Logging, SparkConf}
+      try {
+        if (maxRateLimit > 0) {
+          currentRateLimit.set(newRate.min(maxRateLimit))
+        }
```
we use `} else {` format

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala:
```diff
@@ -34,12 +36,28 @@ import org.apache.spark.{Logging, SparkConf}
-  private lazy val rateLimiter = GuavaRateLimiter.create(desiredRate)
+  // treated as an upper limit
+  private val maxRateLimit = conf.getInt("spark.streaming.receiver.maxRate", 0)
+  private[receiver] var currentRateLimit = new AtomicInteger(maxRateLimit)
```
What is this variable for? Seems like the only place its value is referenced outside this class is RateLimiterSuite. Rather than storing the current rate in the guava RateLimiter as well as this atomicInt, isnt it easier to just keep it one location, that is, the guava rate limiter? If you want to access the rate limit externally, then just have a def that returns, guava rateLimiter.getRate().

dragos committed:
- removed field to hold the current rate limit in rate limiter
- made rate limit a Long and default to Long.MaxValue (consequence of the above)
- removed custom `waitUntil` and replaced it by `eventually`

AmplabJenkins commented:
Merged build triggered.

dragos committed:
Added a few tests that measure the receiver’s rate.

As I mentioned before, I don’t think this is a great idea:

- such tests are flaky (original test in ReceiverSuite was ignored for
that reason)
- Guava’s code has its own test suite, so we can assume it implements
`setRate` correctly

I noticed one flaky failure in about 10 runs on my machine (receiver got
1 message less than the lower bound, which is within 5% of the nominal rate).

dragos commented:
See my last commit message about what changes I made for testing.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37836/consoleFull) for   PR 7471 at commit [`0c51959`](https://github.com/apache/spark/commit/0c51959c9315f63bc80a7ff5b716f48f907b1152).

SparkQA commented:
[Test build #37830 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37830/consoleFull) for   PR 7471 at commit [`61d3ecf`](https://github.com/apache/spark/commit/61d3ecf15acfe228bd6066b15f82983da6962c16).

SparkQA commented:
[Test build #37835 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37835/consoleFull) for   PR 7471 at commit [`7112cfd`](https://github.com/apache/spark/commit/7112cfde8ef265ce49961daff31b9eb276c54f11).

SparkQA commented:
[Test build #37830 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37830/console) for   PR 7471 at commit [`61d3ecf`](https://github.com/apache/spark/commit/61d3ecf15acfe228bd6066b15f82983da6962c16).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37836/console) for   PR 7471 at commit [`0c51959`](https://github.com/apache/spark/commit/0c51959c9315f63bc80a7ff5b716f48f907b1152).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37835 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37835/console) for   PR 7471 at commit [`7112cfd`](https://github.com/apache/spark/commit/7112cfde8ef265ce49961daff31b9eb276c54f11).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala:
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.streaming.receiver
 
+import java.util.concurrent.atomic.AtomicInteger
```
This is not needed any more. Make sure to remove the empty line below too.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/receiver/RateLimiterSuite.scala:
```diff
@@ -0,0 +1,154 @@
+    (blockGenerator, blockGeneratorListener)
+  }
+
+  test("throttling block generator") {
```
This is a good attempt, but this is not the right attempt. This is a RateLimiterSuite, so ideally it should not be using something like a BlockGenerator to test the RateLimiter. So I would update the first three tests ("rate limiter **") to use a dummy rate limiter tester, that tests the upper limit that has been set, before and after any rate update has been made. The tester can be a simple integer increment. Note that only the upper limit needs to be tested, unlike the flaky block generator tests which test for both upper and lower limit.

tdas replied:
Ignore this. See comment above.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala:
```diff
@@ -72,15 +77,50 @@ class ReceiverTrackerSuite extends TestSuiteBase {
 private class DummyReceiver(host: Option[String] = None)
   extends Receiver[Int](StorageLevel.MEMORY_ONLY) {
 
+  var started = false
```
This needs to be @volatile, otherwise waiting threads (one calling `eventually`) may never see the changed value ---> flakiness

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala:
```diff
@@ -72,15 +77,50 @@ class ReceiverTrackerSuite extends TestSuiteBase {
+ * We need the receiver to be an object, otherwise serialization will create another one
+ * and we won't be able to read its rate limit.
+ */
+private object TestDummyReceiver extends DummyReceiver
```
I said this in another comment thread, but since that thread is hidden let me put it here as well. 

You could use reflection to get the private method "executor". Just make the DummyReceiver class have a method that gets the underlying supervisor using reflection. Then, instead of another class/object TestDummyReceiver, you can just use `object DummyReceiver extends DummyReceiver`

tdas replied:
Why did you not remove TestDummyReceiver? You could only make a `object DummyReceiver extends DummyReceiver` and rename `TestReceiverInputDStream` to `DummyReceiverInputDStream`. No need to introduce another `Test***` (Test and Dummy is really redundant).

dragos replied:
Is your issue only about naming? I personally find it weird to have the companion object of a class extend the class itself. It's true that I *could* do that, but I prefer not to. If you feel strongly about it I can do the renaming you wish, but I find it unnecessary.

tdas replied:
Then let's call it SingletonReceiver. I find using test and dummy both in
the name very superfluous

On Wednesday, July 22, 2015, Iulian Dragos <notifications@github.com> wrote:

> In
> streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala
> <https://github.com/apache/spark/pull/7471#discussion_r35208599>:
>
> > +/** An input DStream with a hard-coded receiver that gives access to internals for testing. */
> > +private class TestReceiverInputDStream(@transient ssc_ : StreamingContext)
> > +  extends ReceiverInputDStream[Int](ssc_) {
> > +
> > +  override def getReceiver(): DummyReceiver = TestDummyReceiver
> > +
> > +  def getCurrentRateLimit: Option[Long] = {
> > +    TestDummyReceiver.executor.getCurrentRateLimit
> > +  }
> > +}
> > +
> > +/**
> > + * We need the receiver to be an object, otherwise serialization will create another one
> > + * and we won't be able to read its rate limit.
> > + */
> > +private object TestDummyReceiver extends DummyReceiver
>
> Is your issue only about naming? I personally find it weird to have the
> companion object of a class extend the class itself. It's true that I
> *could* do that, but I prefer not to. If you feel strongly about it I can
> do the renaming you wish, but I find it unnecessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7471/files#r35208599>.
>


-- 
Sent from Gmail Mobile

tdas replied:
Then let's call it SingletonReceiver. I find using test and dummy both in
the name very superfl

On Wednesday, July 22, 2015, Iulian Dragos <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> In
> streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala
> <https://github.com/apache/spark/pull/7471#discussion_r35208599>:
>
> > +/** An input DStream with a hard-coded receiver that gives access to internals for testing. */
> > +private class TestReceiverInputDStream(@transient ssc_ : StreamingContext)
> > +  extends ReceiverInputDStream[Int](ssc_) {
> > +
> > +  override def getReceiver(): DummyReceiver = TestDummyReceiver
> > +
> > +  def getCurrentRateLimit: Option[Long] = {
> > +    TestDummyReceiver.executor.getCurrentRateLimit
> > +  }
> > +}
> > +
> > +/**
> > + * We need the receiver to be an object, otherwise serialization will create another one
> > + * and we won't be able to read its rate limit.
> > + */
> > +private object TestDummyReceiver extends DummyReceiver
>
> Is your issue only about naming? I personally find it weird to have the
> companion object of a class extend the class itself. It's true that I
> *could* do that, but I prefer not to. If you feel strongly about it I can
> do the renaming you wish, but I find it unnecessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7471/files#r35208599>.
>


-- 
Sent from Gmail Mobile

tdas commented:
Things to change for the next iteration. 
- Revert block generate tests, and changes to ReceiverTracker (see inline threads for explanation)
- Revert Receiver class and modify DummyReceiver in ReceiverTrackerSuite to access the `executor` field using Java reflection (again, see other inline thread for discussion)

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -180,6 +180,12 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
 
+  /** Update a receiver's maximum ingestion rate */
+  def sendRateUpdate(streamUID: Int, newRate: Long): Unit = {
+    for (info <- receiverInfo.get(streamUID); eP <- Option(info.endpoint))
```
nit: add `{` and `}` around the for-loop body

dragos committed:
Revert "Added a few tests that measure the receiver’s rate."

This reverts commit 0c51959c9315f63bc80a7ff5b716f48f907b1152.

dragos committed:
Use Reflection for accessing truly private `executor` method and
use the listener bus to know when receivers have registered (`onStart`
is called before receivers have registered, leading to flaky behavior).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37950 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37950/consoleFull) for   PR 7471 at commit [`162d9e5`](https://github.com/apache/spark/commit/162d9e598040b5b2bad36fa0f7139a99df95e79d).

SparkQA commented:
[Test build #37950 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37950/console) for   PR 7471 at commit [`162d9e5`](https://github.com/apache/spark/commit/162d9e598040b5b2bad36fa0f7139a99df95e79d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala:
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
   }
+
+  test("Receiver tracker - propagates rate limit") {
+    object streamingListener extends StreamingListener {
```
This should be capitalized ... say.. TestStreamingListener

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala:
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
+    ssc.scheduler.listenerBus.start(ssc.sc)
+
+    val newRateLimit = 100L
+    val ids = new TestReceiverInputDStream(ssc)
```
ids --> dstream. I got confused with ids being IDs

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala:
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
 }
 
+/** An input DStream with a hard-coded receiver that gives access to internals for testing. */
+private class TestReceiverInputDStream(@transient ssc_ : StreamingContext)
```
TestReceiverInputDStream --> DummyReceiverInputDStream

dragos replied:
I'll find a better name. IMO `Dummy` means nothing, so if names are important we could probably find a better one here. This isn't a dummy input dstream at all, it's a giving access to its rate limit.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala:
```diff
@@ -72,8 +78,62 @@ class ReceiverTrackerSuite extends TestSuiteBase {
+}
+
+/**
+ * We need the receiver to be an object, otherwise serialization will create another one
```
nit: make it more scala docs style to begin with. "A receiver as an object, so that we can read its rate limit. Otherwise,.."

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala:
```diff
@@ -34,12 +34,26 @@ import org.apache.spark.{Logging, SparkConf}
+  def getCurrentLimit: Long =
+    rateLimiter.getRate.toLong
+
+  private[receiver] def updateRate(newRate: Long): Unit =
```
nit: Need a scala doc, for consistency

dragos committed:
Renames and other nitpicks.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38081 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38081/consoleFull) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).

SparkQA commented:
[Test build #38081 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38081/console) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

dragos commented:
Spurious failure:

========================================================================
Building Spark
========================================================================
[info] Building Spark (w/Hive 0.13.1) using SBT with these arguments:  -Pyarn -Phadoop-2.3 -Dhadoop.version=2.3.0 -Pkinesis-asl -Phive-thriftserver -Phive package assembly/assembly streaming-kafka-assembly/assembly streaming-flume-assembly/assembly
Using /usr/java/latest as default JAVA_HOME.
Note, this will be overridden by -java-home if it is set.
**Error: Invalid or corrupt jarfile build/sbt-launch-0.13.7.jar**

SparkQA commented:
[Test build #1157 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1157/consoleFull) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).

SparkQA commented:
[Test build #1157 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1157/console) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

dragos commented:
Same error :cry:

tdas commented:
Jenkins, test this please

SparkQA commented:
[Test build #1175 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1175/consoleFull) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).

SparkQA commented:
[Test build #1176 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1176/consoleFull) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).

SparkQA commented:
[Test build #1175 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1175/console) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1177 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1177/consoleFull) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).

SparkQA commented:
[Test build #1177 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1177/console) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1176 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1176/console) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1179 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1179/consoleFull) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).

SparkQA commented:
[Test build #1179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1179/console) for   PR 7471 at commit [`8941cf9`](https://github.com/apache/spark/commit/8941cf91b03aa7835a78bc756bee1f32cb7bb1d8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

tdas commented:
LGTM. Merging this
